Finish setup of add-on v0.1.0

acorncom opened the pull request:
- [x] delete `index.js`
- [x] upgrade to latest Denali version
- [x] check on typeMap changes

acorncom committed:
Applied latest blueprints

acorncom committed:
Added in node-orm2 to our package.json file

acorncom committed:
Tweaking type mapping to sync up properly

acorncom commented:
@davewasmer This is ready for a review.  I have a bit more to do to get record serialization to work, but not sure that's an issue with this package (may be denali)

knownasilya commented on package.json:
```diff
@@ -11,14 +11,20 @@
     "denali-addon",
     "addon"
   ],
+  "main": "app/addon.js",
```
dist/app/addon.js

acorncom replied:
Doesn't work if dist is there, as dist is also being appended in addon.js

knownasilya replied:
https://github.com/denali-js/denali/blob/master/lib/cli/blueprints/addon/files/__name__/package.json#L13

knownasilya replied:
Also `index.js` should be deleted as well.

davewasmer replied:
FWIW - I would love to be able to get rid of references to `dist/`, as you might have noticed by my `// TODO` comments in a couple places. It's annoying that the runtime code needs to be aware of the build system.

acorncom replied:
@knownasilya Sounds good on the `index.js` file (and thanks, I missed that on the blueprint update).

Regarding `dist` being there in the blueprint, what I ran into when running this with the `ember-help-wanted-backend` app is that the addition of `dist` (in `dist/app/addon.js`) combined with https://github.com/denali-js/denali/blob/master/lib/runtime/addon.js#L161 meant that the addon couldn't be loaded by Denali.  Hence the change 😀 

So how do we want to handle this?  Unless I set up the app wrong (a possible scenario in all of this) the double dist path that Denali looks for (which would match blueprints) would lead to in inoperable add-on 😀

davewasmer replied:
@acorncom see the conversation on denali-js/denali#97 for details

knownasilya commented on package.json:
```diff
@@ -11,14 +11,20 @@
     "denali": "davewasmer/denali",
+    "eslint": "^2.11.0",
+    "eslint-config-denali": "^1.1.0",
+    "denali": "0.0.6",
```
You could change this to 0.0.10

acorncom replied:
Looks like I actually missed changing the line above too ;-)

davewasmer commented on app/orm-adapters/node-orm2.js:
```diff
@@ -7,7 +7,7 @@ import snakeCase from 'lodash/snakeCase';
 
 const typeMap = {
-  text: String,
+  string: String,
```
This type map is meant to correspond to [node-orm2's data types](https://github.com/dresende/node-orm2/wiki/Model%20Properties#types), which uses 'text' for any string value.

acorncom replied:
@davewasmer Understood and sounds good.  The reason I changed this here though was that the add-on was failing with errors about a type not being found. Once this line was changed to `string`, I stopped getting errors (and seemed to have a functioning orm). Will check into this further ...

acorncom replied:
@davewasmer Just checked this again.  With my change above, at this point I'm getting JSON-API responses from the backend from a Sqlite DB! Without it, I get this error from `node-orm2`:

```
ERROR - AssertionError: "string" data type is not supported by the node-orm2 adapter. Supported types are: text, number, boolean, json, date
```

Those docs must be out of date, but the error above seems pretty clear.

acorncom committed:
Shifted back to a dist/ setup in our package.json file

acorncom committed:
Updated denali dep

acorncom committed:
Reverted back to text as issues were caused by my model definition

acorncom committed:
feat(data): shifted to match orm adapter changes upstream

See https://github.com/denali-js/denali/pull/116/files for more details

acorncom committed:
Removed a console.log

acorncom committed:
feat(data): map between Denali camelCase attr names and underscored column names

acorncom commented on app/orm-adapters/node-orm2.js:
```diff
@@ -7,11 +7,11 @@ import snakeCase from 'lodash/snakeCase';
+  number: 'number',
+  boolean: 'boolean',
+  json: 'object',
+  date: 'date'
```
Left side is the ORM name to expose to our Denali model.  Right side is the corresponding ORM name to use when talking to our actual ORM.  Had to change away from actual types due to the use of `mapsTo` below.

acorncom commented on app/orm-adapters/node-orm2.js:
```diff
@@ -39,66 +39,68 @@ export default class NodeORM2Adapter extends ORMAdapter {
+        attributes[key] = {
+          mapsTo: this.serializeKey(key),
+          type: this.typeForAttribute(attribute)
+        }
```
@davewasmer This is what you were after, correct?

davewasmer replied:
Hm, looking at this again now, I think I was actually mistaken in my approach. If node-orm2 types are exposed through to the Denali models, theres no reason to map them to different values. You should just supply whatever type node-orm2 expects on your Denali model, and this adapter can just pass through. Although perhaps I'm missing something - it's late :wink:.

👍  on the `mapsTo` part.

acorncom replied:
This was why I was puzzled originally back when we discussed this a few days ago.  Seemed like what I was seeing with your mapping idea would allow us to by-and-large setup some "Denali model type" defaults that could be mapped within each ORM adapter.  We could allow ORM adapters to expose additional, ORM-specific types in addition if desired, but could allow for some portability between ORMs that way.

I'm fine with not going that route if you'd rather not, but this mapping was what threw me off when looking at it previously ;-)

acorncom commented:
@davewasmer This is working fine on my end now that I've dodged the bug mentioned here (https://github.com/denali-js/denali/issues/127).  Believe we're ready to merge this! 😀 

I'm happy to send over docs on how to use this as well once we've got it merged (or on this PR if you'd rather)

davewasmer commented on config/initializers/node-orm2.js:
```diff
@@ -19,9 +19,6 @@ export default {
         let adapterModels = mapValues(models, (Model) => {
-          if (Model.hasOwnProperty('abstract') && Model.abstract) {
-            return;
-          }
```
I think the goal here was to avoid creating a node-orm2 model for ApplicationModel

knownasilya replied:
Ah that needs a comment then 👍

acorncom replied:
@davewasmer Yep and I liked the idea.  However in https://github.com/denali-js/denali/commit/59ec7e0adc237ec7b7b26eb00a0a9e98ab3e8573 you removed it, so was trying to follow your lead there ... 😀 In the future do you want me questioning your actions a little more loudly? 😉

davewasmer replied:
Ah, yes, that commit message was not very clear. I removed abstract from the default model blueprint because it should be an adapter specific option. So this adapter could support it while others don't.

And by all means, question loudly 😉

acorncom replied:
Sounds good.  Re: questioning loudly, will do.  Starting now ... 😉 

I would argue that having abstract models is a good thing (and important enough that it should be a default blueprint option) that should be available regardless.  It wouldn't have to be used all the time, but I think it adds needless complexity to require that to be re-implemented on a per ORM-adapter level ...  I'd argue strongly for reverting https://github.com/denali-js/denali/commit/59ec7e0adc237ec7b7b26eb00a0a9e98ab3e8573 as a result ... 😄

davewasmer replied:
@acorncom made convincing arguments in Gitter for `abstract` being part of the Denali model API. 👍  to include it. Of course, it will be up to each ORM adapter to figure out how to implement abstract base classes sanely against their ORM.

acorncom committed:
Shifted back to leaving out abstract models

acorncom commented on config/initializers/node-orm2.js:
```diff
@@ -18,12 +18,15 @@ export default {
-        let adapterModels = mapValues(models, (Model) => {
+
+        let adapterModels = {};
+        forEach(models, (Model, key) => {
```
I changed from `mapValues` to `forEach` here because `mapValues` was leaving an `undefined` value in the `adapterModels` hash which then broke `adapter.defineRelationships` below ...

acorncom committed:
Added basic usage instructions and info on how to use it

acorncom committed:
Upgraded to latest Denali addon blueprints

acorncom committed:
chore: add yarn lockfile

acorncom committed:
Add a troubleshooting tip related to using the sqlite driver

Turns out that was the cause of this problem:
https://github.com/denali-js/denali/issues/157

acorncom committed:
Merge branch 'setting-up-further' of github.com:acorncom/denali-node-orm2 into setting-up-further

acorncom committed:
Updated to the latest addon blueprints

acorncom committed:
Fixed two small bugs caught by Yarn

acorncom committed:
Updated the yarn lock file

acorncom committed:
Merge branch 'setting-up-further' of github.com:acorncom/denali-node-orm2 into setting-up-further

acorncom committed:
Merge branch 'setting-up-further' of github.com:acorncom/denali-node-orm2 into setting-up-further

acorncom committed:
Updated to the latest addon blueprints

acorncom committed:
Fixing eslint errors

acorncom committed:
Updated to use the latest Denali release

acorncom committed:
Fixed some bugs introduced when fixing eslint issues

acorncom committed:
Removed typeMap setup

We won't use this any more

acorncom committed:
Tweaked docs regarding types supported

acorncom committed:
Updated Denali dependency to float

acorncom commented:
@davewasmer We should be ready to merge this at this point, it's working well on my end.

davewasmer commented:
Sorry that took so long @acorncom! Thanks for hanging in there ;)

acorncom commented:
@davewasmer No problem, if you couldn't tell, I haven't had much OSS time recently ;-)
